### PR TITLE
Data frame

### PR DESCRIPTION
**Pull request by dantleech**
Introduces the concept of a data frame.

The data frame can be used programatically in reports and can be accessed through the expression language. The data frame is introduced as a new parameter `suite` in the expression report.

This addition will enable any value from the suite to be referenced, the following expression f.e. will show a % comparison to the named benchmark for each row:

```
percent_diff(mode(result_time_avg), mode(suite[subject_name = "benchFoo"]["result_time_avg"]))
```

TODO:

- [x] Create DataFrame and associated classes
- [x] Support expression access for DataFrames 
- [x] Improve tests
- [x] Regression test perforamance
- [x] Update docs
- [x] Split out unrelated improvements/features

## Performance

Initially report generation performance was degraded due to the way the evaluation cache worked (it serialized the parameters as a cache key), removing the cache actually results in a significant gain for report generation at the expense of assertions and progtress loggers:

```
+------------------------------+---------------+-------------+------+-----+------------------+-------------------+----------------+
| benchmark                    | subject       | set         | revs | its | mem_peak         | mode              | rstdev         |
+------------------------------+---------------+-------------+------+-----+------------------+-------------------+----------------+
| ParserBench                  | benchEvaluate | comp. w/tol | 10   | 10  | 1.883mb -2.83%   | 74.146μs +1.40%   | ±1.18% -62.76% |
| ParserBench                  | benchEvaluate | comp.       | 10   | 10  | 1.883mb -2.83%   | 63.677μs +0.55%   | ±2.23% -6.48%  |
| AssertionProcessorBench      | benchAssert   | 0           | 10   | 10  | 3.093mb -3.53%   | 412.578μs +29.59% | ±1.16% -24.57% |
| ExpressionGeneratorBench     | benchGenerate | 0           | 10   | 10  | 12.214mb -52.33% | 8.260ms -35.83%   | ±1.51% -20.55% |
| RunCommandBench              | benchDefault  | 0           | 1    | 10  | 5.881mb -1.02%   | 73.072ms -20.05%  | ±2.25% -3.37%  |
| RunCommandBench              | benchInBand   | 0           | 1    | 10  | 5.881mb -1.02%   | 74.573ms -25.33%  | ±1.45% -50.43% |
| RunCommandBench              | benchNoEnv    | 0           | 1    | 10  | 5.501mb -1.26%   | 21.621ms -3.46%   | ±2.16% -10.81% |
| VariantSummaryFormatterBench | benchFormat   | 0           | 10   | 10  | 3.269mb +1.98%   | 622.163μs +9.24%  | ±2.13% +6.92%  |
+------------------------------+---------------+-------------+------+-----+------------------+-------------------+----------------+
```

As the regressions are at are sub-millisecond it seems like an acceptable trade-off with the added benefit of removing a cache.

### CODE REVIEW

File: lib/Console/Command/Handler/RunnerHandler.php [hunk 1]
```diff
@@ -83,7 +83,7 @@ public function __construct(
 
     public static function configure(Command $command): void
     {
-        $command->addArgument(self::ARG_PATH, InputArgument::OPTIONAL, 'Path to benchmark(s)');
+        $command->addArgument(self::ARG_PATH, InputArgument::OPTIONAL | InputArgument::IS_ARRAY, 'Path to benchmark(s)');
```
Comment (dantleech) [inline lib/Console/Command/Handler/RunnerHandler.php:86]: This is a new "feature" should extract to a new PR

File: lib/Expression/Func/FirstFunction.php [hunk 2]
```diff
@@ -14,7 +14,7 @@ public function __invoke(ListNode $list): Node
         $first = reset($values);
 
         if (!$first) {
-            throw new EvaluationError($list, 'List is empty, cannot get first');
+            return new NullNode();
```
Comment (dantleech) [inline lib/Expression/Func/FirstFunction.php:17]: This is a small change in behavior, should be an own PR

File: lib/Data/DataFrame.php [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    private $serieses;
+
+    /**
+     * @var array
+     */
+    private $columns;
+
+    public function __construct(array $rows, array $columns)
+    {
+        $this->serieses = $rows;
```
Comment (dantleech) [inline lib/Data/DataFrame.php:30]: decide what these are called.

File: lib/Data/DataFrame.php [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+    {
+        $this->series = $series;
+        $this->columns = array_values($columns);
+    }
+
+    /**
+     * @param Series[] $rows
+     * @param string[] $columns
+     */
+    public static function fromRowArray(array $rows, array $columns): self
```
Comment (dantleech) [inline lib/Data/DataFrame.php:38]: row arrays

File: lib/Data/DataFrame.php [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+
+    /**
+     * @var array
+     */
+    private $columns;
+
+    public function __construct(array $series, array $columns)
+    {
+        $this->series = $series;
+        $this->columns = array_values($columns);
```
Comment (dantleech) [inline lib/Data/DataFrame.php:31]: Could add some validation here - ensure that the series are valid

File: lib/Data/DataFrame.php [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+    {
+        return new self(array_map(function (array $row) {
+            return new Series($row);
+        }, $rows), $columns);
+    }
+
+    /**
+     * @param array<int, array<string,mixed>> $records
+     */
+    public static function fromRecords(array $records): self
```
Comment (dantleech) [inline lib/Data/DataFrame.php:57]: => fromRows

File: lib/Data/Series.php [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        $this->values = $values;
+    }
+
+    public function toValues(): array
+    {
+        return $this->values;
+    }
+
+    /**
+     * @return mixed
```
Comment (dantleech) [inline lib/Data/Series.php:26]: change to `int|float|string`

File: lib/Data/DataFrame.php (hunk 6 above)
Comment (dantleech) [reply lib/Data/DataFrame.php:57]: There is a difference - rows do not have keys

File: lib/Expression/NodeEvaluator/ParameterEvaluator.php [hunk 8]
```diff
@@ -2,12 +2,18 @@
 
 namespace PhpBench\Expression\NodeEvaluator;
 
+use ArrayAccess;
```
Comment (dantleech) [inline lib/Expression/NodeEvaluator/ParameterEvaluator.php:5]: Rename => ParameterAccessEvaluator

File: lib/Expression/Ast/PropertyAccessNode.php [hunk 9]
```diff
@@ -2,10 +2,10 @@
 
 namespace PhpBench\Expression\Ast;
 
-class ParameterNode implements Node
+class PropertyAccessNode implements Node
```
Comment (dantleech) [inline lib/Expression/Ast/PropertyAccessNode.php:5]: this is a b/c break for themes

File: lib/Expression/Lexer.php [hunk 10]
```diff
@@ -14,11 +14,11 @@ final class Lexer
      */
     private $unitNames;
 
-    private const PATTERN_NAME = '(?:[a-z_\/]+)';
+    private const PATTERN_NAME = '(?:[a-z_\/]{1}[a-z0-9_\/]*)';
```
Comment (dantleech) [inline lib/Expression/Lexer.php:17]: this change allows numbers in variable names (test?)

File: lib/Expression/Lexer.php [hunk 11]
```diff
@@ -14,11 +14,11 @@ final class Lexer
...
     private const PATTERN_FUNCTION = '(?:[a-z_\/]+\()';
     private const PATTERN_STRING = '"(?:[^"]|"")*"';
     private const PATTERN_NUMBER = '(?:[0-9]+(?:[\\.][0-9]+)*)(?:e[+-]?[0-9]+)?';
-    private const PATTERN_PARAMETER = '(?:(?:[a-z_]+\\.)+(?:[a-z_]+))';
+    private const PATTERN_PARAMETER = '(?:(?:[a-z_]+\\.)+(?:[a-z_\[\]]+))';
```
Comment (dantleech) [inline lib/Expression/Lexer.php:21]: this allow square brackets in parameter/variable names

File: lib/Data/DataFrame.php [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+                $value = $series->value($index);
+                $values[$name][] = $value;
+            }
+        }
+
+        return $values;
+    }
+
+    /**
+     * @return array<string, array<mixed>>
```
Comment (dantleech) [inline lib/Data/DataFrame.php:184]: return `scalar`

File: lib/Data/DataFrame.php [hunk 13]
```diff
@@ -0,0 +1,233 @@
...
+    {
+        return in_array($offset, $this->columns);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public function offsetGet($offset)
+    {
+        return $this->column($offset)->toValues();
```
Comment (dantleech) [inline lib/Data/DataFrame.php:208]: tests for array access?

File: lib/Report/Generator/ExpressionGenerator.php [hunk 14]
```diff
@@ -86,7 +87,7 @@ public function configure(OptionsResolver $options): void
     %s, 
     coalesce(
         first(subject_time_unit),
-        "microseconds"
+        "time"
```
Comment (dantleech) [inline lib/Report/Generator/ExpressionGenerator.php:90]: This is a separate fix (add to changelog)

File: lib/Report/Generator/ExpressionGenerator.php [hunk 15]
```diff
@@ -207,22 +217,39 @@ private function aggregate(array $table, array $aggregateCols): array
...
 
-            foreach (($row[SuiteCollectionTransformer::COL_HAS_BASELINE][0] ? array_merge($exprMap, $baselineExprMap) : $exprMap) as $name => $expr) {
+            $exprMap = $frame->row(0)->get(SuiteCollectionTransformer::COL_HAS_BASELINE) ? array_merge(
+                $exprMap,
+                $baselineExprMap
+            ) : $exprMap;
+
+            $columnValues = array_merge(
+                [
+                    'suite' => $allFrame
```
Comment (dantleech) [inline lib/Report/Generator/ExpressionGenerator.php:238]: worth prefixing `_`? the columns are are not extensible however

File: lib/Report/Transform/SuiteCollectionTransformer.php [hunk 16]
```diff
@@ -13,12 +14,9 @@ final class SuiteCollectionTransformer
 {
     public const COL_HAS_BASELINE = 'has_baseline';
 
-    /**
-     * @return array<string,array<string,mixed>>
-     */
-    public function suiteToTable(SuiteCollection $collection, bool $includeBaseline = false): array
+    public function suiteToTable(SuiteCollection $collection, bool $includeBaseline = false): DataFrame
```
Comment (dantleech) [inline lib/Report/Transform/SuiteCollectionTransformer.php:17]: this is a BC break potentially - add another method and deprecate the `array` one

**Comment by tillkruss** [comment]
This is fantastic!

Screen Shot 2021-05-16 at 5 17 14 PM

**Comment by dantleech** [comment]
Nice! I'm not sure percent_diff works here though - red is "bad", while negative is "good". You could also change the expresion to use the `x` format

**Comment by tillkruss** [comment]
Agreed. The red is a little aggressive. Could you explain how to do the `Nx` notation?

**Comment by dantleech** [comment]
Math and concatenation: `(mode(suite[subjectName="foo"]) / mode(result_time_avg)) ~ "x"` or something ...
